src/lib/cloud-provider.js

```diff
@@ -0,0 +1,124 @@
+import log from './log.js';
+
+
+class CloudProvider {
+    constructor (cloudHost, vm, user, projectId) {
+        this.vm = vm;
+        this.username = user;
+        this.projectId = projectId;
+
+        // Open a websocket connection to the clouddata server
+        this.openConnection(cloudHost);
+    }
+
+    /**
+     * Open a new websocket connection to the clouddata server.
+     * @param {string} cloudHost The cloud data server to connect to.
+     */
+    openConnection (cloudHost) {
+        if (window.WebSocket === null) {
+            log.warn('Websocket support is not available in this browser');
+            this.connection = null;
+            return;
+        }
+
+        this.connection = new WebSocket((location.protocol === 'http:' ? 'ws://' : 'wss://') + cloudHost);
+
+        this.connection.onerror = e => {
+            log.error(`Websocket connection error: ${JSON.stringify(e)}`);
+
+            // TODO Add re-connection attempt logic here
+        };
+
+        this.connection.onmessage = event => {
+            const messageString = event.data;
+            log.info(`Received websocket message: ${messageString}`);
+            const message = JSON.parse(messageString);
+            if (message.method === 'set') {
+                const varData = {
+                    projectId: this.projectId,
+                    varUpdate: {
+                        projectId: message.project_id,
```

Comment:
A comment here to explain the two different `projectId`'s would be nice.

Reply:
Oops! That was a remnant of old code that I decided wasn't necessary anymore.